tmk_core/protocol/chibios/usb_main.c

```diff
@@ -827,24 +827,7 @@ uint8_t keyboard_leds(void) {
 }
 
 void send_report(uint8_t endpoint, void *report, size_t size) {
-    osalSysLock();
-    if (usbGetDriverStateI(&USB_DRIVER) != USB_ACTIVE) {
-        osalSysUnlock();
-        return;
-    }
-
-    if (usbGetTransmitStatusI(&USB_DRIVER, endpoint)) {
-        /* Need to either suspend, or loop and call unlock/lock during
-         * every iteration - otherwise the system will remain locked,
-         * no interrupts served, so USB not going through as well.
-         * Note: for suspend, need USB_USE_WAIT == TRUE in halconf.h */
-        if (osalThreadSuspendTimeoutS(&(&USB_DRIVER)->epc[endpoint]->in_state->thread, TIME_MS2I(10)) == MSG_TIMEOUT) {
-            osalSysUnlock();
-            return;
-        }
-    }
-    usbStartTransmitI(&USB_DRIVER, endpoint, report, size);
-    osalSysUnlock();
+    usbTransmit(&USB_DRIVER, endpoint, report, size);
```

Comment:
This change effectively makes `send_report()` use infinite timeout — if the host does not poll the corresponding input endpoint, `usbTransmit()` will block forever when called the second time.

One way to reproduce this is using a slightly modified `joystick` keymap for onekey:
```diff
--- a/keyboards/handwired/onekey/keymaps/joystick/keymap.c
+++ b/keyboards/handwired/onekey/keymaps/joystick/keymap.c
@@ -5,7 +5,7 @@
 #endif

 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
-    LAYOUT_ortho_1x1(JS_0)
+    LAYOUT_ortho_1x1(KC_A)
 };

 void matrix_scan_user(void) {
```
and then using
```sh
qmk flash -kb handwired/onekey/rp2040 -km joystick -e MOUSEKEY_ENABLE=no -e EXTRAKEY_ENABLE=no
```
(for some reason `JOYSTICK_SHARED_EP=yes` and `DIGITIZER_SHARED_EP=yes` are effectively forced if the shared endpoint is ever used, so everything except the joystick is disabled instead).

With this configuration at least in Linux the key sending `KC_A` does not work unless some app actually reads the joystick events (e.g., `jstest-gtk`, and the properties window for the joystick needs to be open).

Maybe something similar could happen when trying to use a more usual keyboard configuration in BIOS, if any event gets sent over the shared endpoint for some reason (the host might not be polling the shared endpoint in such limited environment).

Reply:
Ah, that's ... fun. 

Though, it seems that this block is what seems to be causing issues:
```c
    if (usbGetTransmitStatusI(&USB_DRIVER, endpoint)) {
        /* Need to either suspend, or loop and call unlock/lock during
         * every iteration - otherwise the system will remain locked,
         * no interrupts served, so USB not going through as well.
         * Note: for suspend, need USB_USE_WAIT == TRUE in halconf.h */
        if (osalThreadSuspendTimeoutS(&(&USB_DRIVER)->epc[endpoint]->in_state->thread, TIME_MS2I(10)) == MSG_TIMEOUT) {
            osalSysUnlock();
            return;
        }
    }
```
And I'm not really sure the best way to fix that.

Reply:
My proposal for a fix is https://github.com/qmk/qmk_firmware/pull/21656 😉